src/lif_cell_group.hpp

```diff
@@ -1,220 +1,293 @@
+                last_time_updated_.resize(cells.size());
+
+                for (const auto& cell : cells) {
+                    cells_.push_back(util::any_cast<lif_cell_description>(cell));
+                }
+
+                // Initialize variables for the external Poisson input.
+                for (auto lid : util::make_span(0, cells_.size())) {
+                    EXPECTS(cells_[lid].n_poiss >= 0);
+                    EXPECTS(cells_[lid].w_poiss >= 0);
+                    EXPECTS(cells_[lid].d_poiss >= 0);
+                    EXPECTS(cells_[lid].rate >= 0);
+
+                    // If a cell receives some external Poisson input then initialize the corresponding variables.
+                    if (cells_[lid].n_poiss > 0) {
+                        lambda_.push_back(1.0/(cells_[lid].rate * cells_[lid].n_poiss));
+                        generator_[lid].seed(1000 + first_gid + lid);
+                        sample_next_poisson(lid);
+                    }
+                }
+            }
+
+            cell_kind get_cell_kind() const override {
+                return cell_kind::lif_neuron;
+            }
+
+            // TODO: Remove this after testing.
+            // Samples voltage from time t to tfinal with resolution
+            // given by class parameter sampling_dt_
+            void sample_voltage(value_type v, value_type E_L, value_type tau_m, time_type tend, bool refractory_period) {
+                time_type start = last_time_voltage_updated_;
+
+                for (time_type time = last_time_voltage_updated_; time < tend; time += sampling_dt_) {
+                    if (!refractory_period) {
+                        v *= exp(-(time - last_time_voltage_updated_)/tau_m);
+                    }
+                    else {
+                        // If in refractory period, then at the moment of spike,
+                        // let v has original value (which is >= V_thr) and
+                        // otherwise, let v be equal to the resting potential E_L.
+                        if (time > start) {
+                            v = E_L;
+                        }
+                    }
+                    // Put a pair of time and voltage.
+                    voltage_.push_back({time, v});
+                    // Update last time the voltage is updated.  last_time_voltage_updated_ = time;
+                }
+            }
+
+            void advance(time_type tfinal, time_type dt) override {
+                PE("lif");
+                // Distribute incoming events to individual cells.
+                while (!events_.empty()) {
+                    // Put the event to the corresponding cell's queue.
+                    auto ev = events_.front();
+                    events_.pop();
+
+                    int target_gid = ev.target.gid;
+                    // Transform gid -> lid.
+                    int lid = target_gid - gid_base_;
+
+                    cell_events_[lid].push(ev);
```

Comment:
This could be optimised by pushing events directly to the per-cell queues `cell_events_` in `::enqueue_events()`.